File: BalancedBinaryTree/first.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
        if(helper(root)==-1) return false;
        else return true;
    }
    private int helper(TreeNode root){
        if(root==null) return 0;
        int leftH=helper(root.left);
        if(leftH==-1 ) return -1;
        int rightH=helper(root.right);
        if( rightH ==-1) return -1;

        if(Math.abs(leftH-rightH)>1) return -1;

        return 1+Math.max(leftH,rightH);

    }
}

File: BalancedBinaryTree/first.cpp
class Solution {
public:
bool isBalanced(TreeNode* root) {
    solution1(root); 
    return chk;
    //return solution2(root, height); 
}
private:
bool chk = true;
int height = 0;
int solution1(TreeNode *root){
    if(root == NULL) return 0;
    int left = solution1(root->left);
    int right = 0;
    if(chk){
        right = solution1(root->right);
        if(abs(left - right) > 1) chk = false;
    }
    return max(left, right) + 1;
}
bool solution2(TreeNode *root, int &height){
    if(!root){
        height = 0;
        return true;
    }
    int left, right;
    if(!solution2(root->left, left) || !solution2(root->right, right) || abs(left - right) > 1) return false;
    height = max(left, right) + 1;
    return true;
}
};
